src/main/java/dev/tr7zw/entityculling/occlusionculling/OcclusionCullingInstance.java

```diff
@@ -82,132 +188,157 @@ private boolean isVisible(Vec3d start, Vec3d[] targets) {
-																	// the first cell
+				n += y - (int) MathUtil.floor(y1); // increment total amount of intersecting cells
+				t_next_y = (float) ((y0 - MathUtil.floor(y0)) * dt_dy); // calculate the next vertical intersection
+																			// point
+				// based on the position inside
+				// the first cell
 			}
-	
+
 			if (dz == 0f) {
 				z_inc = 0;
 				t_next_z = dt_dz; // don't increment vertically because the Vec3d is perfectly horizontal
 			} else if (z1 > z0) {
 				z_inc = 1; // target point is vertically greater than starting point so increment every
 							// step by 1
-				n += (int) Math.floor(z1) - z; // increment total amount of intersecting cells
-				t_next_z = (float) ((Math.floor(z0) + 1 - z0) * dt_dz); // calculate the next vertical intersection point based on the position inside
-																		// the first cell
+				n += (int) MathUtil.floor(z1) - z; // increment total amount of intersecting cells
+				t_next_z = (float) ((MathUtil.floor(z0) + 1 - z0) * dt_dz); // calculate the next vertical
+																				// intersection
+				// point based on the position inside
+				// the first cell
 			} else {
 				z_inc = -1; // target point is vertically smaller than starting point so reduce every step
 							// by 1
-				n += z - (int) Math.floor(z1); // increment total amount of intersecting cells
-				t_next_z = (float) ((z0 - Math.floor(z0)) * dt_dz); // calculate the next vertical intersection point based on the position inside
-																	// the first cell
+				n += z - (int) MathUtil.floor(z1); // increment total amount of intersecting cells
+				t_next_z = (float) ((z0 - MathUtil.floor(z0)) * dt_dz); // calculate the next vertical intersection
+																			// point
+				// based on the position inside
+				// the first cell
 			}
-	
-			boolean finished =stepRay(start, x0, y0, z0, x, y, z, dt_dx, dt_dy, dt_dz, n, x_inc, y_inc, z_inc, t_next_y, t_next_x,
-					t_next_z);
-			if(finished) {
+
+			boolean finished = stepRay(start, x, y, z, x, y, z, dt_dx, dt_dy, dt_dz, n, x_inc, y_inc, z_inc,
+					t_next_y, t_next_x, t_next_z);
+			if (finished) {
+				cacheResult(targets[0], true);
 				return true;
 			}
 		}
+		cacheResult(targets[0], false);
 		return false;
 	}
-	
-	private boolean stepRay(Vec3d start, double x0, double y0, double z0, int x, int y,
-			int z, double dt_dx, double dt_dy, double dt_dz, int n, int x_inc, int y_inc, int z_inc, double t_next_y,
-			double t_next_x, double t_next_z) {
+
+	private void cacheResult(Vec3d vector, boolean result) {
+		int cx = (int) MathUtil.fastFloor(vector.x + reach);
+		int cy = (int) MathUtil.fastFloor(vector.y + reach);
+		int cz = (int) MathUtil.fastFloor(vector.z + reach);
+		int keyPos = cx + cy * (reach * 2) + cz * (reach * 2) * (reach * 2);
+		int entry = keyPos / 4;
+		int offset = (keyPos % 4) * 2;
+		if (result) {
+			cache[entry] |= 1 << offset;
+		} else {
+			cache[entry] |= 1 << offset + 1;
```

Comment:
I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [BSHIFT: Possible bad parsing of shift operation (BSHIFT_WRONG_ADD_PRIORITY)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bshift-possible-bad-parsing-of-shift-operation-bshift-wrong-add-priority).
The code performs an operation like (x << 8 + y). Although this might be correct, probably it was meant to perform (x << 8) + y, but shift operation has a lower precedence, so it's actually parsed as x << (8 + y).